docs/designs/allreduce.md

```diff
@@ -1,14 +1,198 @@
-# Design for Allreduce Support
+# Design for Allreduce-based Training Support
 
-This document describes the design for supporting Allreduce in ElasticDL. Note that this is still a work-in-progress.
+This document describes the design for supporting allreduce-based distributed training in ElasticDL.
+
+## Motivation
+
+While distributed training based on parameter servers can support training very large models and datasets by adding
+more workers and parameter servers as described in [parameter server design doc](parameter_server.md), there are additional
+challenges involved in order to optimize the performance, for example:
+
+* It is not easy to identify the right ratio of the number of workers to the number of parameter servers. For example,
+if only a small number of parameter servers are used, there will likely be network communication overheads. If many parameter
+servers are used, the communication may saturate network interconnects.
+* If the model could fit within the computational resources on each worker, additional maintenance and communication overheads are
+introduced when the model is partitioned to multiple parameter servers.
+* Additional computational resources for replicating models of each parameter server are needed to support fault-tolerance.
+
+In contrast, distributed training based on collective communication primitives such as [allreduce](https://mpitutorial.com/tutorials/mpi-reduce-and-allreduce/)
+and [ring-allreduce](http://research.baidu.com/bringing-hpc-techniques-deep-learning/) could be more efficient and easier
+to use in certain use cases. There are many existing technologies available that provide implementations for these collective
+communication primitives and please head over to the section on [existing collective communication technologies](#existing-collective-communication-technologies)
+for details if interested. Allreduce-based distributed training could address many of the challenges mentioned above, for example:
+
+* Each worker stores a complete set of model parameters. In other words, no parameter server is needed so it's straightforward
+to add more workers when necessary.
+* Failures among the workers can be recovered easily by restarting the failed workers and then load the current model sent from
+any of the existing workers. Model does not need to be replicated to support fault-tolerance, which avoids the waste of resources.
+* The model can be updated more efficiently by fully leveraging the network structure and collective communication algorithms.
+For example, in ring-allreduce algorithm, each of *N* workers only needs to communicate with two of its peer workers *2 * (N − 1)* times.
```

Comment:
`2*(N-1) times to get all parameter updates`?

Reply:
Thanks. Edited.